user/tests/wiring/power_management_long_running/power.cpp

```diff
@@ -0,0 +1,379 @@
+
+    SERIAL.println("> Wait the power management settled...");
+    assertTrue(waitFor([]{
+        return System.batteryState() == BATTERY_STATE_NOT_CHARGING;
+    }, 10000));
+    assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+    PMIC power;
+    assertFalse(power.isChargingEnabled());
+
+    SERIAL.println("> Test is running...");
+    time32_t start = Time.now();
+    while ((Time.now() - start) < TEST_DURATION) {
+        assertTrue(System.batteryState() == BATTERY_STATE_NOT_CHARGING);
+        assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+        delay(1s);
+    }
+
+    g_state = CHARGING_STATE_ENABLED_CHARGING;
+}
+
+test(power_03_battery_charging_when_charging_enabled) {
+    SERIAL.println("\r\npower_03_battery_charging_when_charging_enabled");
+    if (g_state != CHARGING_STATE_ENABLED_CHARGING) {
+        skip();
+        return;
+    }
+
+    SERIAL.println("> Enable charging.");
+    int ret = System.setPowerConfiguration(System.getPowerConfiguration().clearFeature(SystemPowerFeature::DISABLE_CHARGING));
+    assertEqual(ret, (int)SYSTEM_ERROR_NONE);
+    auto cfg = System.getPowerConfiguration();
+    assertEqual(cfg.isFeatureSet(SystemPowerFeature::DISABLE_CHARGING), false);
+
+    SERIAL.println("> Wait the power management settled...");
+    assertTrue(waitFor([]{
+        return System.batteryState() == BATTERY_STATE_CHARGING;
+    }, 10000));
+    assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+    PMIC power;
+    assertTrue(power.isChargingEnabled());
+
+    SERIAL.println("> Test is running...");
+    while (System.batteryState() == BATTERY_STATE_CHARGING) {
+        assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+        delay(5s);
+    }
+    SERIAL.println("> Battery is charged.");
+    assertTrue(System.batteryState() == BATTERY_STATE_CHARGED);
+
+    g_state = CHARGED_STATE;
+}
+
+test(power_04_battery_charged) {
+    SERIAL.println("\r\npower_04_battery_charged");
+    if (g_state != CHARGED_STATE) {
+        skip();
+        return;
+    }
+
+    constexpr time32_t TEST_DURATION = 10 * 60; // 10 minutes
+
+    assertTrue(System.batteryState() == BATTERY_STATE_CHARGED);
+    assertTrue(System.powerSource() == POWER_SOURCE_USB_HOST);
+    PMIC power;
+    assertTrue(power.isChargingEnabled());
+
+    // When the battery is charged and the USB is connected, the battery should
+    // reside in the charged state. While for problematic battery, the voltage
+    // drops too much once stop charging due to charged. If the voltage drops more
+    // than 100mV, the PMIC will start charging again. From the human view, the charging
+    // status LED will blink.
+    SERIAL.println("> Test is running...");
+    time32_t start = Time.now();
+    while ((Time.now() - start) < TEST_DURATION) {
+        if (System.batteryState() == BATTERY_STATE_CHARGING) {
+            SERIAL.println("> WARN: (problematic) battery is charging!");
```

Comment:
I had a Boron stuck in this state with approximately 70% of battery charge, which would indicate that we dropped out of https://github.com/particle-iot/device-os/pull/2290/files#diff-92b482059ee678b515c6002f68f3c2ace005b1ecb29490a6ef2e87b72de3a0c4R228 prematurely and the test above anyway succeeded. We should potentially add some debouncing and/or checks for actual SoC. It may also indicate a problem with power manager.

Reply:
I've figured out the potential fake battery state in power manager. I'll add an assertion in the previous test case to check if the SoC complies with the charged state.